chain/network/src/peer_manager/tests/snapshot_hosts.rs

```diff
@@ -0,0 +1,249 @@
+use crate::network_protocol::testonly as data;
+use crate::network_protocol::SnapshotHostInfo;
+use crate::network_protocol::SyncSnapshotHosts;
+use crate::peer;
+use crate::peer_manager;
+use crate::peer_manager::peer_manager_actor::Event as PME;
+use crate::tcp;
+use crate::testonly::{make_rng, AsSet as _};
+use crate::types::NetworkRequests;
+use crate::types::PeerManagerMessageRequest;
+use crate::types::PeerMessage;
+use near_async::time;
+use near_crypto::SecretKey;
+use near_o11y::testonly::init_test_logger;
+use near_o11y::WithSpanContextExt;
+use near_primitives::hash::CryptoHash;
+use near_primitives::network::PeerId;
+use near_primitives::types::EpochHeight;
+use near_primitives::types::ShardId;
+use peer_manager::testonly::FDS_PER_PEER;
+use pretty_assertions::assert_eq;
+use rand::Rng;
+use std::collections::HashSet;
+use std::sync::Arc;
+
+/// Create an instance of SnapshotHostInfo for testing purposes
+fn make_snapshot_host_info(
+    peer_id: &PeerId,
+    secret_key: &SecretKey,
+    rng: &mut impl Rng,
+) -> Arc<SnapshotHostInfo> {
+    let epoch_height: EpochHeight = rng.gen::<EpochHeight>();
+    let shards_num: usize = rng.gen_range(1..16);
+    let shards: Vec<ShardId> = (0..shards_num).map(|_| rng.gen()).collect();
```

Comment:
It shouldn't matter for these tests, but small, distinct values would be more realistic for shard ids. It works like this in practice:
- The shard layout for the protocol has shards labelled `0, 1, ... N-1`. Today `N = 4`.
- A snapshot contains some subset of those shards.

The components you are testing here (network infra and the SnapshotHostsCache) don't really care about these properties and can handle arbitrary values, so I don't necessarily think these tests need to change.

Unless/until we introduce additional verification of these values within the network side of things, the downstream consumers of this content should be able to handle unexpected values for ShardId anyway. @VanBarbascu

Reply:
I intentionally didn't limit the values here because I wanted to see what happens :P
Maybe the cache allocates `vec![0; max_shard_id+1]` and the program would crash? I felt that it would make the test more throughout. After all malicious peers could send messages with large shard values, so it should be handled gracefully.

Reply:
That makes sense.

I'll throw out one other idea (but please feel free to merge this as-is and move on), which would be to conduct the "happy path" tests on valid data and add some separate test`test_unexpected_shard_ids` which passes large shard values to see what happens. Then in the future if we happen to introduce some downstream consumer which does something like you are imagining (e.g. allocate a vec of size `max_shard_id + 1`) the failing test would pinpoint the exact problem.

Reply:
Good idea, I made shard generation more realistic and added a separate test which makes sure that `PeerManager` handles big shard ids correctly.

Reply:
One thing that worries me a bit is that this test doesn't pass when I send the message as the peer manager itself. It looks like a PeerManager doesn't process its own messages. Is this a concern, or is it working as expected?
Here's a failing test, it hangs on the last line:
```
/// Send a SyncSnapshotHosts message with very large shard ids.
/// Makes sure that PeerManager processes large shard ids without any problems.
#[tokio::test]
async fn large_shard_id_in_cache_pure_manager() {
    init_test_logger();
    let mut rng = make_rng(921853233);
    let rng = &mut rng;
    let mut clock = time::FakeClock::default();
    let chain = Arc::new(data::Chain::make(&mut clock, rng, 10));

    tracing::info!(target:"test", "Create a peer manager.");
    let pm = peer_manager::testonly::start(
        clock.clock(),
        near_store::db::TestDB::new(),
        chain.make_config(rng),
        chain.clone(),
    )
    .await;

    tracing::info!(target:"test", "Send a SnapshotHostInfo message with very large shard ids.");
    let big_shard_info = Arc::new(SnapshotHostInfo::new(
        pm.peer_info().id,
        CryptoHash::hash_borsh(1234_u64),
        1234,
        vec![0, 1232232, ShardId::MAX - 1, ShardId::MAX],
        &pm.cfg.node_key,
    ));

    let message = PeerManagerMessageRequest::NetworkRequests(NetworkRequests::SnapshotHostInfo {
        sync_hash: big_shard_info.sync_hash,
        epoch_height: big_shard_info.epoch_height,
        shards: big_shard_info.shards.clone(),
    });

    pm.actix.addr.send(message.with_span_context()).await.unwrap();

    tracing::info!(target:"test", "Make sure that the message is received and processed without any problems.");
    let want: HashSet<Arc<SnapshotHostInfo>> = std::iter::once(big_shard_info).collect();
    pm.wait_for_snapshot_hosts(&want).await;
}
```

Reply:
OK, interesting case. It is technically working as expected because `NetworkRequest::SnapshotHostInfo` triggers a network message to the peers of the node:
https://github.com/near/nearcore/blob/bf7ba1189917bded51a339865469f9e5e06dac0a/chain/network/src/peer_manager/peer_manager_actor.rs#L787-L802

However, it sounds reasonable to also insert the `SnapshotHostInfo` to the local cache here.

Reply:
Should we do something with it? I don't feel qualified to decide, since you're the expert on this code.
Can I merge the PR, or should I go and "fix" this "issue"?

Reply:
I like the idea of inserting to the local cache here (in addition to broadcasting to all peers) if you want to sneak that in with this PR.